src/Microsoft.Extensions.Configuration.AzureAppConfiguration/AzureAppConfigurationProvider.cs

```diff
@@ -356,25 +362,22 @@ private async Task RefreshIndividualKeyValues()
 
                         if (watchedKv != null)
                         {
+                            if (changeWatcher.RefreshAll)
+                            {
+                                shouldRefreshAll = true;
+                                break;
+                            }
+
+                            hasChanged = true;
+
                             // Add the key-value if it is not loaded, or update it if it was loaded with a different label
                             _settings[watchedKey] = watchedKv;
-                            hasChanged = true;
                         }
                     }
 
                     if (hasChanged)
                     {
-                        if (changeWatcher.RefreshAll)
-                        {
-                            shouldRefreshAll = true;
-
-                            // Skip refresh for other key-values since refreshAll will populate configuration from scratch
-                            break;
-                        }
-                        else
-                        {
-                            await SetData(_settings).ConfigureAwait(false);
-                        }
+                        await SetData(_settings).ConfigureAwait(false);

```

Comment:
>await SetData [](start = 24, length = 13)

Not relevant to this PR, but it looks like calling this in a loop could cause us to do a lot of key value processor calls. That could mean a lot of key vault calls. Maybe we could optimize to build a list of KeyValuePair and then update settings all at once instead of multiple calls.

Just not sure how common it is to watch a lot of single key-values.

Reply:
Yes, we can move it outside the loop to call it once instead of once per watcher and reduce the calls to adapters. One downside I can think of that approach is that if we exhaust both retry attempts to load a specific watcher, we lose the settings fetched previously. In order to preserve those settings, we might need some refactoring for the exception logic.